src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_secrets.rs

```diff
@@ -811,3 +814,114 @@ fn index_of(input_string: String, search_char: char) -> i32
+        if detections.is_empty() {
+            return input.to_string();
+        }
+
+        // Merge positions into ranges of characters to replace.
+        let mut current_detections = Vec::new();
+        let mut current_detection : Option<Detection> = None;
+
+        detections.sort_unstable_by_key(|item| item.start);
+
+        for detection in detections.iter() {
+            let scan_match = detection.matches_bytes(input_as_bytes, true).unwrap();
+            let match_text = scan_match.text();
+
+            if validate_checksum {
+                let match_text_as_bytes = match_text.as_bytes();
+                let mut signature_bytes = vec![0; 3];
+                signature_bytes[0] = match_text_as_bytes[57];
+                signature_bytes[1] = match_text_as_bytes[58];
+                signature_bytes[2] = match_text_as_bytes[59];
+
+                let signature = general_purpose::STANDARD.encode(&signature_bytes);
+
+                let checksum_validation_result = try_validate_common_annotated_key(
+                    &match_text,
+                    &signature,
+                );
+
+                assert!(checksum_validation_result);
+            }
+
+            // Get c3id for the find
+            let match_text_c3id = cross_company_correlating_id::generate_cross_company_correlating_id(match_text);
+
+            let redaction_token = match default_redaction_token {
+                Some(token) => token,
+                None => &format!("SEC101/200:{}", match_text_c3id),
+            };
+
+            if current_detection.is_none() {
+                current_detection = Some(Detection {
+                    start: detection.start(),
+                    end: detection.start() + detection.len() as u64,
+                    length: detection.len(),
+                    redaction_token: redaction_token.to_string()
+                });
+                current_detections.push(current_detection.clone().unwrap());
+            }
+            else {
+                if detection.start <= current_detection.clone().unwrap().end {
+                    // Overlapping case or contiguous case.
+                    let current = current_detection.as_mut().unwrap();
+                    current.length = (std::cmp::max(current.end,
+                                                    detection.start() + detection.len() as u64)
+                        - current.start) as usize;
+                } else {
+                    current_detection = Some(Detection {
+                        start: detection.start(),
+                        end: detection.start() + detection.len() as u64,
+                        length: detection.len(),
+                        redaction_token: redaction_token.to_string()
+                    });
+                    current_detections.push(current_detection.clone().unwrap());
+                }
+            }
+        }
+
+        let mut result = String::new();
```

Comment:
It would be nice to have an in-place masker instead of text copying. This will work fine for small lines of text, but not large lines of text without overheads.

Reply:
@beaubelgrave Interesting. So the overhead scenario might be something like: <long blob of text> <1 secret> <another long blob of text>. In this case, with the current implementation, we would needlessly end up copying both the blobs, even though we just needed a single replacement. I guess the C# implementation also suffers from this performance bottleneck. 
Let me take a stab at optimizing the Rust method based on your feedback.

Reply:
OK @beaubelgrave I took a stab at implementing an "in-place" masker. With this, there was a 31% reduction in the masking time. Look forward to your thoughts on this.

@michaelcfanning for awareness. If this implementation is finalized, we may want to modify the C# implementation based along the Rust algorithm.